backend/routeHandlers/studentRouteHandler.js

```diff
@@ -413,13 +413,205 @@ const addStudentEffort = async (personalityTestAttemptKeys, students) => {
     }
 };
 
+const addStudentPreferences = async (personalityTestAttemptKeys, students) => {
+    /**
+     * Implements the logic required to add 'Preferences' personality data to the
+     * database given the unique column requirements it has.
+     */
+    const resultInsertData = [];
+    // format the data so it fulfils the database column requirements
+    personalityTestAttemptKeys.forEach((attempt) => {
+        // find the student who made this attempt
+        const [student] = students.filter((student) => {
+            return student.studentId === attempt.student_id;
+        });
+        // add their data in a suitable format
+        resultInsertData.push([
+            attempt.test_attempt_id,
+            student.submission_timestamp,
+        ])
+    });
+
+    try {
+        await promiseBasedQuery(
+            "INSERT IGNORE INTO preference_submission" +
+            "(personality_test_attempt, submission_timestamp) " +
+            "VALUES ?;",
+            [resultInsertData]
+        );
+
+    } catch (err) {
+        console.log(err);
+    }
+
+    // is this it?
+    // addProjectPreferences(personalityTestAttemptKeys, students)
+};
+
+const addProjectPreferences = async (personalityTestAttemptKey, students) => {
+    /**
+     * Implements the logic for adding each project preference to a given preference submission
+     */
+    const resultInsertData = [];
+    // format the data so it fulfils the database column requirements
+    // For each student 
+    personalityTestAttemptKeys.forEach(async (attempt) => {
+        // find the student who made this attempt
+        const [student] = students.filter((student) => {
+            return student.studentId === attempt.student_id;
+        });
+
+        // get the preference submission ID 
+        let preference_submission_id;
+        try {
+            const [{ preference_submission_id: id }] = await promiseBasedQuery(
+                "SELECT preference_submission_id FROM preference_submission WHERE personality_test_attempt=?;",
+                [attempt.test_attempt_id]
+            );
+            preference_submission_id = id;
+        } catch (err) {
+            console.log(err);
+        }
+
+        // Turning the string preferences into an array of numbers 
+        const preferencesArray = student.preferences.split(" ").map(Number);
+
+        // for each project 
+        for (let i = 0; i < preferencesArray.length; i++) {
+            resultInsertData.push([
+                preference_submission_id,
+                i,
+                preferencesArray[i]
+            ]);
+
+            // insert for each project 
+            // this looks like bad practice to insert in a for loop, im not sure how to fix since there will be N project_preference for N projects
```

Comment:
You can insert multiple rows in sql. If you build a string (or a list and make it a string at the end) and insert it after the for loop... but its finee for now
https://www.sqlservertutorial.net/sql-server-basics/sql-server-insert-multiple-rows/

Reply:
fixed i think